### PR TITLE
fix the bug (total=0 rows) and the corresponding test case.

### PR DESCRIPTION
**Pull request by HavierD**
Now, duplicate rows with total=0 in the middle of the sheet won't be removed. Only bottom duplicate rows with total=0 will be removed.

### CODE REVIEW

File: api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java [hunk 1]
```diff
@@ -230,4 +230,13 @@ void duplicateRowsShouldBeRemoved() throws Exception {
         assertEquals(20, validRows.size());
     }
 
+    @Test
```
Comment (utas-raymondng) [inline api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java:233]: Can you add comment to your test case so that what it is doing?

**Review by utas-raymondng — CHANGES_REQUESTED** [review]

File: api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java (hunk 1 above)
Comment (HavierD) [reply api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java:233]: > Can you add comment to your test case so that what it is doing?

Got it.

**Comment by utas-raymondng** [comment]
Don't merge until nrmn software version upgrade completed

File: api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java [hunk 2]
```diff
@@ -378,4 +381,25 @@ public void downloadFile(final HttpServletResponse response, @PathVariable Strin
                     jobId, clientException.getMessage())));
         }
     }
+
```
Comment (utas-raymondng) [inline api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java:384]: Bad function call that modify pass in object value, please return something instead of modify the existing one

Collection<Long> filterNonBottomDuplicateRows(Collection<Long>..... , Long);

File: api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java [hunk 3]
```diff
@@ -378,4 +381,25 @@ public void downloadFile(final HttpServletResponse response, @PathVariable Strin
...
+    private void filterNonBottomDuplicateRows(Collection<Long> rowIdsToRemove, Long lastRowId) {
+
+        var toRemoveIds = new ArrayList<Long>();
+
+        if (!rowIdsToRemove.contains(lastRowId)) {
```
Comment (utas-raymondng) [inline api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java:389]: Is this a reliable check, can the lastRowId not empty and then it will not exist in the rowIdsToRemove()?

File: api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java [hunk 4]
```diff
@@ -230,4 +230,18 @@ void duplicateRowsShouldBeRemoved() throws Exception {
...
 
+    /** This test is for:
+     * When adding new jobs, only rows (with "total=0" and duplicate more than 3 times) at the end of the sheet should be
+     * removed. Rows (with "total=0" and duplicate more than 3 times) are not at the end of the sheet should be kept.
+     * @throws Exception
+     */
+    @Test
+    void removeBottomDuplicateRowsOnly() throws Exception {
+        FileSystemResource file = new FileSystemResource("src/test/resources/sheets/removeDuplicateBottomRows.xlsx");
+        MockMultipartFile mockFile = new MockMultipartFile("sheets/removeDuplicateBottomRows.xlsx", file.getInputStream());
```
Comment (utas-raymondng) [inline api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java:241]: The removeDuplicateBottomRows contains 495 lines, the last 7 line is total=0 and repeat more then 3 times,  the first two line are header, so 495 - 7 - 2 = 486 so why 487 is expected result?

File: api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java [hunk 5]
```diff
@@ -230,4 +230,18 @@ void duplicateRowsShouldBeRemoved() throws Exception {
...
+        var parsedSheet = sheetService.stageXlsxFile(mockFile, false);
+        var validRows = stagedJobController.getRowsToSave(parsedSheet);
+        assertEquals(487, validRows.size());
+    }
```
Comment (utas-raymondng) [inline api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java:245]: Please add a few edge case, 

1. The xls file last line is "ID = 486"
2. The xls file last line is "ID = 490"
3. Assert shows that line "ID = 464, 465, 466, 467, 468" exist in the validRows

**Review by utas-raymondng — CHANGES_REQUESTED** [review]
Please see comments

File: api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java (hunk 2 above)
Comment (HavierD) [reply api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java:384]: Got it. Will modify soon.

File: api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java (hunk 3 above)
Comment (HavierD) [reply api/src/main/java/au/org/aodn/nrmn/restapi/controller/StagedJobController.java:389]: Sorry I didn't get your point. The meaning of this part is: Because we only want to remove duplicate rows on the bottom. So, if the `rowIdsToRemove` doesn't contain the last row, it means no rows should be removed.

File: api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java (hunk 4 above)
Comment (HavierD) [reply api/src/test/java/au/org/aodn/nrmn/restapi/service/SpreadSheetServiceIT.java:241]: to record after discussion: if the last row is valid (species is "SURVEY NOT DONE", "DEBRIS - ZERO", or "NO SPECIES FOUND"), it will keep this row. Need to add a new test for it (valid, keep it. Otherwise remove it).

**Review by utas-raymondng — APPROVED** [review]
LGTM
